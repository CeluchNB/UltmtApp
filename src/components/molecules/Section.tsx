import * as React from 'react'
import { useColors } from '../../hooks'
import { Button, IconButton } from 'react-native-paper'
import { FlatList, ListRenderItem, StyleSheet, Text, View } from 'react-native'
import { size, weight } from '../../theme/fonts'

interface SectionProps {
    title: string
    onButtonPress: () => void
    numColumns?: number
    listData: any[]
    renderItem: ListRenderItem<any>
    buttonText: string
    error?: string
    showCreateButton?: boolean
    onCreatePress?: () => void
}

const Section: React.FC<SectionProps> = ({
    title,
    onButtonPress,
    numColumns = 1,
    listData,
    renderItem,
    buttonText,
    error,
    showCreateButton = false,
    onCreatePress = () => {},
}) => {
    const { colors } = useColors()

    const styles = StyleSheet.create({
        container: {
            width: '100%',
        },
        titleContainer: {
            display: 'flex',
            flexDirection: 'row',
        },
        title: {
            fontSize: size.fontLarge,
            color: colors.textPrimary,
            fontWeight: weight.bold,
            marginBottom: 10,
            flex: 1,
        },
        createButton: {
            display: !showCreateButton ? 'none' : 'flex',
            flexWrap: 'nowrap',
            backgroundColor: colors.textPrimary,
            borderRadius: 5,
        },
        list: {},
        button: {
            alignSelf: 'center',
            marginTop: 10,
        },
        error: {
            fontSize: size.fontMedium,
            marginBottom: 50,
        },
    })

    return (
        <View style={styles.container}>
            <View style={styles.titleContainer}>
                <Text style={styles.title}>{title}</Text>
                <IconButton
                    color={colors.primary}
                    style={styles.createButton}
                    onPress={onCreatePress}
                    icon="plus"
                />
            </View>
            {error ? (
                <Text style={styles.error}>{error}</Text>
            ) : (
                <FlatList
                    listKey={title}
                    data={listData}
                    scrollEnabled={false}
                    numColumns={numColumns}
                    renderItem={renderItem}
                />
            )}
            {!error && (
                <Button
                    mode="text"
                    style={styles.button}
                    color={colors.textPrimary}
                    onPress={onButtonPress}>
                    {buttonText}
                </Button>
            )}
        </View>
    )
}

export default Section
