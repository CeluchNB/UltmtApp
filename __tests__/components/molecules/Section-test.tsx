import * as React from 'react'
import Section from '../../../src/components/molecules/Section'
import renderer from 'react-test-renderer'
import { Text, View } from 'react-native'
import { fireEvent, render } from '@testing-library/react-native'

it('matches snapshot', () => {
    const snapshot = renderer
        .create(
            <Section
                title="title"
                onButtonPress={() => ({})}
                listData={[]}
                renderItem={() => <View />}
                buttonText="button"
                showButton={true}
            />,
        )
        .toJSON()

    expect(snapshot).toMatchSnapshot()
})

it('basic list data displayed', () => {
    const mockFn = jest.fn()

    const { getByText } = render(
        <Section
            title="title"
            onButtonPress={mockFn}
            listData={['item1', 'item2', 'item3']}
            renderItem={({ item }) => {
                return <Text>{item}</Text>
            }}
            buttonText="button"
            showButton={true}
        />,
    )

    expect(getByText('title')).toBeTruthy()
    expect(getByText('item1')).toBeTruthy()
    expect(getByText('item2')).toBeTruthy()
    expect(getByText('item3')).toBeTruthy()
    expect(() => {
        getByText('item4')
    }).toThrow()

    fireEvent.press(getByText('button'))
    expect(mockFn).toHaveBeenCalledTimes(1)
})

it('display with error', () => {
    const mockFn = jest.fn()

    const { getByText } = render(
        <Section
            title="title"
            onButtonPress={mockFn}
            listData={['item1', 'item2', 'item3']}
            renderItem={({ item }) => {
                return <Text>{item}</Text>
            }}
            buttonText="button"
            showButton={false}
            error="No valid data"
        />,
    )

    expect(getByText('No valid data')).toBeTruthy()
    expect(() => {
        getByText('item1')
    }).toThrow()

    expect(() => {
        getByText('button')
    }).toThrow()
})
