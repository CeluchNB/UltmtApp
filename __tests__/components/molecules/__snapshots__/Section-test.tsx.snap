// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<View
  style={
    Object {
      "flex": 0,
      "flexShrink": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "display": "flex",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#ffd600",
          "flex": 1,
          "fontSize": 61.64383561643835,
          "fontWeight": "bold",
          "marginBottom": 10,
        }
      }
    >
      title
    </Text>
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            Object {
              "opacity": 0.32,
            },
            Object {
              "backgroundColor": "#ffd600",
              "borderRadius": 5,
              "display": "none",
              "flexWrap": "nowrap",
            },
          ],
        ]
      }
      testID="create-button"
    >
      <View>
        <Text
          accessibilityElementsHidden={true}
          allowFontScaling={false}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          selectable={false}
          style={
            Array [
              Object {
                "color": "#303030",
                "fontSize": 24,
              },
              Array [
                Object {
                  "lineHeight": 24,
                  "transform": Array [
                    Object {
                      "scaleX": 1,
                    },
                  ],
                },
                Object {
                  "backgroundColor": "transparent",
                },
              ],
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          Û∞êï
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    listKey="title"
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEnabled={false}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
  <View
    collapsable={false}
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "transparent",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0,
        "elevation": 0,
        "marginTop": 10,
        "minWidth": 64,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#ffd600",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          button
        </Text>
      </View>
    </View>
  </View>
</View>
`;
